docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,765 @@
+workflow has followed the "intended path" - to completion of the whole graph,
+say, vs going down a dead-end side branch. This is because all task outputs 
+are effectively *optional*, and the scheduler just follows the flow wherever it
+leads at runtime.
+
+Cylc 7 could tell (albeit imperfectly!) if the intended path was followed 
+because all task outputs were effectively *required*: all paths were
+pre-spawned (to the next cycle point instance of each active task, roughly
+speaking) and users had to explicitly remove unused paths with suicide
+triggers, so any remaining unsatisfied waiting tasks would indicate a problem
+by stalling the workflow.
+
+This document proposes a way to convey the intended path of execution to the
+Cylc 8 spawn-on-demand scheduler by making all task outputs **required** by
+default and giving a new syntax to explicitly mark **optional** outputs. If a
+task finishes without completing all of its required outputs it will be
+retained in the n=0 task pool and flagged as an **incomplete task** (often
+this will coincide with unexpected task failure; otherwise it implies that the
+task did not behave as the workflow configuration expected despite reporting
+success).
+
+*Note: optional triggers were considered and rejected as another way of
+potentially handling this problem - see cylc-admin#128*
+
+## Output Syntax
+
+- `foo:x` means `x` is a *required output* of task `foo`
+  - if a task finishes without completing any required outputs, retain it in
+    the task pool as an *incomplete task*
+
+- `foo:y?` means `y` is an *optional output* of task `foo`
+  - it is OK for a task to finish without completing optional outputs
+
+Abbreviated syntax for the default success case:
+- `foo` means `foo:succeed` (success is required)
+- `foo?` means `foo:succeed?` (success is optional)
+
+### Caveats
+
+An output can't be both required and optional.
+
+Success and failure are mutually exclusive, so if one is optional they must
+both be optional. Optional branches can handle intermittent failure, e.g:
+- a platform that isn't always available: `foo:submit-fail?`
+- a task that you expect to fail sometimes: `foo:fail?`
+
+Failure can't be required. Expecting a task to fail every time and treating
+it as incomplete if it succeeds would essentially amount to reversing normal 
+success and failure symantics. If you really have an executable that returns
+non-`0` whilst still doing what you need in the workflow, provide a wrapper
+that correctly returns success status.
+- `foo:fail  # ERROR`
+- `foo:submit-fail  # ERROR`
+
+Finally, optional `start` doesn't really make sense. At best it would be
+equivalent to `foo:submit?`, so use that instead.
+- `foo:start?  # ERROR`
+
+If a task's success is not referenced anywhere in the graph (i.e. only other
+outputs are referenced) then by definition other tasks do not depend on its
+success, and we can therefore safely assume that success is optional. E.g.:
```

Comment:
What about `submit`? I think we should assume `submit` is required unless it is explicitly made optional.

Reply:
Yeah. OK, changing this for `succeed` too, to make for fewer exceptions.  i.e. if `foo:succeed` or `foo:fail` not mentioned anywhere in the graph (only `foo:x` perhaps) assume that success is required.